src/ess/reduce/ui.py

```diff
@@ -18,145 +20,242 @@
+            tooltip='Clear Output',
         )
-        or widget.enabled
-    }
 
+        def run_workflow(_: widgets.Button) -> None:
+            self.output.clear_output()
+            with self.output:
+                compute_result = workflow_runner()
+                if result_registry is not None:
+                    result_registry.clear()
+                    result_registry.update(compute_result)
+                for i in compute_result.values():
+                    display.display(display.HTML(i._repr_html_()))
+
+        def clear_output(_: widgets.Button) -> None:
+            self.output.clear_output()
+
+        self.run_button.on_click(run_workflow)
+        output_clear_button.on_click(clear_output)
+        button_box = widgets.HBox([self.run_button, output_clear_button])
+        super().__init__([button_box, self.output], **kwargs)
+
+
+def _workflow_runner_template(
+    workflow_constructor: Callable[[], sl.Pipeline],
+    output_selection_box: OutputSelectionWidget,
+    input_selection_box: ParameterBox,
+) -> dict[type, Any]:
+    """Template function that constructs a workflow from inputs of widgets."""
+    workflow = workflow_constructor()
+    target_outputs = output_selection_box.value
+    values = input_selection_box.collect_values()
+    return assign_parameter_values(workflow, values).compute(
+        target_outputs, scheduler=sl.scheduler.NaiveScheduler()
+    )
+
+
+def connect_refresh_button(
+    refresh_button: widgets.Button, output_widget: widgets.Output
+) -> None:
+    def refresh_output(_: widgets.Button):
+        output_widget.clear_output()
+
+    refresh_button.on_click(refresh_output)
+
+
+def connect_output_selection_and_parameter_run_button(
+    *output_selection_widgets: widgets.Widget,
+    parameter_refresh_button: widgets.Button,
+    run_button: widgets.Button,
+) -> None:
+    # Disable run button when output selection changes
+    def observe_selection_change(_) -> None:
+        run_button.disabled = True
+        run_button.tooltip = 'To run the workflow, refresh parameters.'
+
+    for output_selection_widget in output_selection_widgets:
+        output_selection_widget.observe(observe_selection_change)
+
+    # Enable run button when parameters are generated
+    original_run_button_tooltip = run_button.tooltip
+
+    def observe_parameter_refrheshed(_) -> None:
+        run_button.disabled = False
+        run_button.tooltip = original_run_button_tooltip
+
+    parameter_refresh_button.on_click(observe_parameter_refrheshed)
+
+
+class WorkflowWidget(widgets.TwoByTwoLayout):
+    def __init__(
+        self, workflow: sl.Pipeline, result_registry: dict | None = None, **kwargs
+    ):
+        self.output_selection_box = OutputSelectionWidget(workflow)
+        registry_getter = partial(
+            _registry_getter_template, workflow, self.output_selection_box
+        )
+        self.parameter_box = ParameterBox(registry_getter)
+        workflow_runner = partial(
```

Comment:
(minor) Since we are not using the `_workflow_runner_template` anywhere else, maybe it's better to create the callback function here in the constructor directly?

Like this:

```python
class WorkflowWidget(widgets.TwoByTwoLayout):
    def __init__(
    ):
        ...

        def workflow_runner():
            # We don't have to copy the workflow here because that's already done in assign_parameter_values
            target_outputs = self.output_selection_box.value
            values = self.input_selection_box.collect_values()
            return assign_parameter_values(workflow, values).compute(
                target_outputs, scheduler=sl.scheduler.NaiveScheduler()
            )
        self.result_box = ResultBox(workflow_runner, result_registry)

            ...
```

The advantage is that it removes one layer of indirection and makes it a bit easier to understand.

Same goes for `_registry_getter_template`.

Reply:
Yeah, it used to be like this since it didn't have any wrapper class and it was just a ``Box`` instead.
Also I thought maybe each widgets can be used as it is... which is very unlikely I think...
I'll fix it.